src/bvals/boundary_conditions.hpp

```diff
@@ -14,15 +14,45 @@
 #ifndef BVALS_BOUNDARY_CONDITIONS_HPP_
 #define BVALS_BOUNDARY_CONDITIONS_HPP_
 
+#include <functional>
 #include <memory>
+#include <string>
 
 #include "basic_types.hpp"
 #include "interface/meshblock_data.hpp"
+#include "mesh/domain.hpp"
 
 namespace parthenon {
 
-TaskStatus ApplyBoundaryConditions(std::shared_ptr<MeshBlockData<Real>> &rc);
+// Physical boundary conditions
 
+using BValFunc = std::function<void(std::shared_ptr<MeshBlockData<Real>> &, bool)>;
+
+TaskStatus ProlongateBoundaries(std::shared_ptr<MeshBlockData<Real>> &rc);
+
+TaskStatus ApplyBoundaryConditionsOnCoarseOrFine(std::shared_ptr<MeshBlockData<Real>> &rc,
+                                                 bool coarse);
+
+inline TaskStatus ApplyBoundaryConditions(std::shared_ptr<MeshBlockData<Real>> &rc) {
+  return ApplyBoundaryConditionsOnCoarseOrFine(rc, false);
+}
+
+namespace BoundaryFunction {
+
+void OutflowInnerX1(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void OutflowOuterX1(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void OutflowInnerX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void OutflowOuterX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void OutflowInnerX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void OutflowOuterX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectInnerX1(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectOuterX1(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectInnerX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectOuterX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectInnerX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
+void ReflectOuterX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse);
```

Comment:
Can you make coarse const?

Reply:
In principal, yes. Since these functions are often expected to be supplied by downstream codes via `ApplicationInput`, though... I think I'd rather let `coarse` be non-const though, just to simplify the function signature. I'm willing to change it if you feel strongly.

Reply:
It's just a suggestion, best practice says make it a const if you can.

Reply:
Cool. I'll make the others constant but pass on this one.